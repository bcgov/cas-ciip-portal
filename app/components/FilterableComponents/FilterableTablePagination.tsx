import React from 'react';
import {useRouter} from 'next/router';
import {Pagination} from 'react-bootstrap';

interface Props {
  totalCount: number;
  maxResultsPerPage: number;
}

export const FilterableTablePaginationComponent: React.FunctionComponent<Props> = (
  props
) => {
  const {totalCount, maxResultsPerPage} = props;
  const router = useRouter();
  const maxPages = Math.ceil(totalCount / maxResultsPerPage);

  // Set the activePage by query string value. If no value exists, the activePage is the first page
  let activePage = 1;
  if (router?.query?.pageVars)
    activePage =
      JSON.parse(router?.query?.pageVars?.toString())?.activePage || 1;

  const items = [];

  const paginate = (pageNumber: number) => {
    const paginationObject = {
      offset: (pageNumber - 1) * maxResultsPerPage,
      activePage: pageNumber
    };
    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        pageVars: JSON.stringify(paginationObject)
      }
    };
    router.push(url, url, {shallow: true});
  };

  // Determines what page numbers to render based on the location of the activePage
  let startPage;
  let endPage;
  if (maxPages <= 9) {
    startPage = 1;
    endPage = maxPages;
  } else if (maxPages - activePage <= 4) {
    startPage = maxPages - 8;
    endPage = maxPages;
  } else if (activePage <= 5) {
    startPage = 1;
    endPage = 9;
  } else {
    startPage = activePage - 4;
    endPage = activePage + 4;
  }

  // Populate the Pagination items with page numbers & functionality
  for (let pageNumber = startPage; pageNumber <= endPage; pageNumber++) {
    items.push(
      <Pagination.Item
        key={pageNumber}
        active={pageNumber === activePage}
        onClick={() => paginate(pageNumber)}
      >
        {pageNumber}
      </Pagination.Item>
    );
  }
  if (maxPages > 1)
    return (
      <Pagination>
        <Pagination.First onClick={() => paginate(1)} />
        <Pagination.Prev
          onClick={() => (activePage > 1 ? paginate(activePage - 1) : null)}
        />
        {startPage !== 1 && <Pagination.Ellipsis />}
        {items}
        {endPage !== maxPages && <Pagination.Ellipsis />}
        <Pagination.Next
          onClick={() =>
            activePage !== maxPages ? paginate(activePage + 1) : null
          }
        />
        <Pagination.Last onClick={() => paginate(maxPages)} />
      </Pagination>
    );
  return null;
};

export default FilterableTablePaginationComponent;
