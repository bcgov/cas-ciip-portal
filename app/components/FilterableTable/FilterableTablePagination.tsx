import React from "react";
import { Pagination, Dropdown } from "react-bootstrap";

interface Props {
  /**
   * The total number of items in all of the pages
   */
  totalCount: number;
  /**
   * Defaults to DEFAULT_PAGE_SIZE.
   */
  pageSize?: number;
  /**
   * The number of items to skip to get to the current page. Defaults to 0
   */
  offset?: number;
  onOffsetChange: (offset: number) => void;
  onPageSizeChange: (pageSize: number) => void;
}

export const DEFAULT_PAGE_SIZE = 20;

const PAGE_SIZE_OPTIONS = [20, 50, 100];

const FilterableTablePagination: React.FunctionComponent<Props> = ({
  totalCount,
  pageSize = DEFAULT_PAGE_SIZE,
  offset = 0,
  onOffsetChange,
  onPageSizeChange,
}) => {
  const activePage = Math.floor(offset / pageSize + 1) || 1;
  const maxPages = Math.ceil(totalCount / pageSize);

  const items = [];

  // Determines what page numbers to render based on the location of the activePage
  let startPage;
  let endPage;
  if (maxPages <= 9) {
    startPage = 1;
    endPage = maxPages;
  } else if (maxPages - activePage <= 4) {
    startPage = maxPages - 8;
    endPage = maxPages;
  } else if (activePage <= 5) {
    startPage = 1;
    endPage = 9;
  } else {
    startPage = activePage - 4;
    endPage = activePage + 4;
  }

  const handlePageChange = (pageNumber: number) => {
    onOffsetChange((pageNumber - 1) * pageSize);
  };

  // Populate the Pagination items with page numbers & functionality
  for (let pageNumber = startPage; pageNumber <= endPage; pageNumber++) {
    items.push(
      <Pagination.Item
        key={pageNumber}
        active={pageNumber === activePage}
        onClick={() => handlePageChange(pageNumber)}
      >
        {pageNumber}
      </Pagination.Item>
    );
  }
  return (
    <>
      <div className="pagination">
        <Dropdown>
          <Dropdown.Toggle variant="outline-dark" id="dropdown-basic">
            Items Per Page: <strong>{pageSize}</strong>
          </Dropdown.Toggle>

          <Dropdown.Menu>
            {PAGE_SIZE_OPTIONS.map((option) => (
              <Dropdown.Item
                key={option}
                onClick={() => onPageSizeChange(option)}
              >
                {option}
              </Dropdown.Item>
            ))}
          </Dropdown.Menu>
        </Dropdown>
        {maxPages > 1 && (
          <Pagination style={{ marginLeft: "15px" }}>
            <Pagination.First onClick={() => handlePageChange(1)}>
              &lt;&lt; First Page
            </Pagination.First>
            <Pagination.Prev
              onClick={() =>
                activePage > 1 ? handlePageChange(activePage - 1) : null
              }
            />
            {startPage !== 1 && <Pagination.Ellipsis />}
            {items}
            {endPage !== maxPages && <Pagination.Ellipsis />}
            <Pagination.Next
              onClick={() =>
                activePage !== maxPages
                  ? handlePageChange(activePage + 1)
                  : null
              }
            />
            <Pagination.Last onClick={() => handlePageChange(maxPages)}>
              Last Page &gt;&gt;
            </Pagination.Last>
          </Pagination>
        )}
      </div>
      <style jsx>{`
        .pagination {
          display: flex;
        }
        :global(.pagination a.page-link) {
          color: #036;
        }
        :global(.pagination .page-item.active .page-link) {
          background-color: #036;
          border-color: #036;
        }
      `}</style>
    </>
  );
};

export default FilterableTablePagination;
