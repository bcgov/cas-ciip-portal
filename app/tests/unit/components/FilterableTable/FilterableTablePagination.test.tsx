import React from "react";
import { shallow } from "enzyme";
import FilterableTablePagination from "components/FilterableTable/FilterableTablePagination";

describe("PaginationBar", () => {
  it("should not render the Pagination component if Math.ceil(totalCount / pageSize) < 2", async () => {
    const r = shallow(
      <FilterableTablePagination
        totalCount={5}
        onOffsetChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />
    );
    expect(r.find("PageItem").exists()).toBe(false);
    expect(r).toMatchSnapshot();
  });
  it("should render the Pagination component if Math.ceil(totalCount / pageSize) > 1", async () => {
    const r = shallow(
      <FilterableTablePagination
        totalCount={21}
        onOffsetChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />
    );
    expect(r.find("PageItem").exists()).toBe(true);
    expect(r).toMatchSnapshot();
  });
  it("should create a number of pagination buttons equal to Math.ceil(totalCount / pageSize)", async () => {
    const r = shallow(
      <FilterableTablePagination
        totalCount={45}
        onOffsetChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />
    );
    expect(r.find("PageItem").at(0).text()).toBe("1");
    expect(r.find("PageItem").at(1).text()).toBe("2");
    expect(r.find("PageItem").at(2).text()).toBe("3");
    expect(r.find("PageItem").at(3).exists()).toBe(false);
  });
  it("should render a maximum of 9 pagination pages if Math.ceil(totalCount / pageSize) > 9", async () => {
    const r = shallow(
      <FilterableTablePagination
        totalCount={190}
        onOffsetChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />
    );
    expect(r.find("PageItem").at(8).text()).toBe("9");
    expect(r.find("PageItem").at(9).exists()).toBe(false);
    expect(r.find("Ellipsis").exists()).toBe(true);
    expect(r).toMatchSnapshot();
  });
  it("should set the first item in the pagination bar to active by default", async () => {
    const r = shallow(
      <FilterableTablePagination
        totalCount={130}
        onOffsetChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />
    );
    expect(r.find("PageItem").at(0).prop("active")).toBe(true);
  });
});
