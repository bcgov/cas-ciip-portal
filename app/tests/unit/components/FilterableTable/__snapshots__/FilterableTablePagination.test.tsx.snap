// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationBar should not render the Pagination component if Math.ceil(totalCount / pageSize) < 2 1`] = `
<Fragment>
  <div
    className="jsx-1384315916 pagination"
  >
    <Dropdown
      navbar={false}
    >
      <DropdownToggle
        id="dropdown-basic"
        variant="outline-dark"
      >
        Items Per Page: 
        <strong
          className="jsx-1384315916"
        >
          20
        </strong>
      </DropdownToggle>
      <DropdownMenu
        align="left"
        alignRight={false}
        flip={true}
      >
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="20"
          onClick={[Function]}
        >
          20
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="50"
          onClick={[Function]}
        >
          50
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="100"
          onClick={[Function]}
        >
          100
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  </div>
  <JSXStyle
    id="1384315916"
  >
    .pagination.jsx-1384315916{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.pagination a.page-link{color:#036;}.pagination .page-item.active .page-link{background-color:#036;border-color:#036;}
  </JSXStyle>
</Fragment>
`;

exports[`PaginationBar should render a maximum of 9 pagination pages if Math.ceil(totalCount / pageSize) > 9 1`] = `
<Fragment>
  <div
    className="jsx-1384315916 pagination"
  >
    <Dropdown
      navbar={false}
    >
      <DropdownToggle
        id="dropdown-basic"
        variant="outline-dark"
      >
        Items Per Page: 
        <strong
          className="jsx-1384315916"
        >
          20
        </strong>
      </DropdownToggle>
      <DropdownMenu
        align="left"
        alignRight={false}
        flip={true}
      >
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="20"
          onClick={[Function]}
        >
          20
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="50"
          onClick={[Function]}
        >
          50
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="100"
          onClick={[Function]}
        >
          100
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
    <ForwardRef
      style={
        Object {
          "marginLeft": "15px",
        }
      }
    >
      <First
        onClick={[Function]}
      >
        &lt;&lt; First Page
      </First>
      <Prev
        onClick={[Function]}
      />
      <PageItem
        active={true}
        activeLabel="(current)"
        disabled={false}
        key="1"
        onClick={[Function]}
      >
        1
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="2"
        onClick={[Function]}
      >
        2
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="3"
        onClick={[Function]}
      >
        3
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="4"
        onClick={[Function]}
      >
        4
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="5"
        onClick={[Function]}
      >
        5
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="6"
        onClick={[Function]}
      >
        6
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="7"
        onClick={[Function]}
      >
        7
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="8"
        onClick={[Function]}
      >
        8
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="9"
        onClick={[Function]}
      >
        9
      </PageItem>
      <Ellipsis />
      <Next
        onClick={[Function]}
      />
      <Last
        onClick={[Function]}
      >
        Last Page &gt;&gt;
      </Last>
    </ForwardRef>
  </div>
  <JSXStyle
    id="1384315916"
  >
    .pagination.jsx-1384315916{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.pagination a.page-link{color:#036;}.pagination .page-item.active .page-link{background-color:#036;border-color:#036;}
  </JSXStyle>
</Fragment>
`;

exports[`PaginationBar should render the Pagination component if Math.ceil(totalCount / pageSize) > 1 1`] = `
<Fragment>
  <div
    className="jsx-1384315916 pagination"
  >
    <Dropdown
      navbar={false}
    >
      <DropdownToggle
        id="dropdown-basic"
        variant="outline-dark"
      >
        Items Per Page: 
        <strong
          className="jsx-1384315916"
        >
          20
        </strong>
      </DropdownToggle>
      <DropdownMenu
        align="left"
        alignRight={false}
        flip={true}
      >
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="20"
          onClick={[Function]}
        >
          20
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="50"
          onClick={[Function]}
        >
          50
        </DropdownItem>
        <DropdownItem
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          disabled={false}
          key="100"
          onClick={[Function]}
        >
          100
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
    <ForwardRef
      style={
        Object {
          "marginLeft": "15px",
        }
      }
    >
      <First
        onClick={[Function]}
      >
        &lt;&lt; First Page
      </First>
      <Prev
        onClick={[Function]}
      />
      <PageItem
        active={true}
        activeLabel="(current)"
        disabled={false}
        key="1"
        onClick={[Function]}
      >
        1
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="2"
        onClick={[Function]}
      >
        2
      </PageItem>
      <Next
        onClick={[Function]}
      />
      <Last
        onClick={[Function]}
      >
        Last Page &gt;&gt;
      </Last>
    </ForwardRef>
  </div>
  <JSXStyle
    id="1384315916"
  >
    .pagination.jsx-1384315916{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.pagination a.page-link{color:#036;}.pagination .page-item.active .page-link{background-color:#036;border-color:#036;}
  </JSXStyle>
</Fragment>
`;
